Submission/spack blog (#64)
ROCm/rocm-blogs: blogs/software-tools-optimization/spack-installation/README.md

ROCm offically supports roughly 60 components. The exact list of component varies across multiple sources. Here are a few:

### Lists of ROCm Components
- [What is ROCm?](https://rocm.docs.amd.com/en/latest/what-is-rocm.html) $^{[1]}$
- [`ROCm/default.xml`](https://github.com/ROCm/ROCm/blob/develop/default.xml) $^{[2]}$
- [ROCm Documentation for Spack](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/how-to/spack.html#rocm-packages-in-spack) $^{[1]}$
- [ROCm Compatibility Matrix](https://rocm.docs.amd.com/en/latest/compatibility/compatibility-matrix.html) $^{[1]}$
- [Version release notes](https://rocm.docs.amd.com/en/latest/about/release-notes.html#rocm-components) $^{[1]}$

$^{[1]}$ Pages from ROCm Documentation

$^{[2]}$ ROCm GitHub repository

The landscape of ROCm components is best laid out by the following graphic, taken from the [What is ROCm?](https://rocm.docs.amd.com/en/latest/what-is-rocm.html) page from the documentation.

<img src="./images/rocm-software-stack-6_3_2.jpg">

<p style="text-align:center">
Figure 1: The ROCm software stack
</p>

> The ROCm stack relies on lower layers of runtimes and compilers which are generally essential for most ROCm components; i.e., these are the "meat and potatoes" of ROCm. Above these layers are **Tools**, which are quite handy for developers, but are also used by some ROCm components. The ROCm stack becomes particularly discrete at the **Library** level. E.g., `rocJPEG` does not depend on `hipTensor` and vice versa. This motivates the question, "Why should we install all of ROCm?" if we are just 

Reviewer: This sentence seems truncated.

Author: Thanks, reworded in latest commit.

### ROCm Component Dependencies

> Spack is useful as a from-source package manager; as such, it also serves as a great system for understanding the dependencies of the ROCm stack. The following graphic has been generated using the outputs of `spack spec <package>`.

Reviewer: Have you seen https://github.com/ROCm/ROCm/blob/develop/tools/rocm-build/ROCm.mk
This is the dependency matrix in Makefile form used by the internal CI.

Author: I had not seen this, this is definitely useful for handling dependencies within the ROCm stack. I suppose the advantages that Spack brings are that 1. the verbosity of `spack spec` provides clarity of dependencies beyond the ROCm stack and 2. this allows one to derive ROCm dependencies for non-ROCm software, e.g. PyTorch. I have no doubt a similar graphic could be generated from the Makefile, but then connections to external packages are lost.

Regardless, I have reworded this section to better reflect why Spack is useful in this case.

Reviewer 2: Is the ROCm.mk file something that is really digestible by a typical ROCm user to understand dependencies between ROCm packages ? While it seems like a great tool for internal CI, what about end users who are asking about how to build ROCm from source ?

Reviewer: There are published instructions for building ROCm from source, in the same way that the internal CI does it. Essentially (i.e. I am skipping over some details, arguements to commands ... ) the steps are

1) repo init # Pull a manifest describing the commits which make up ROCm.
2) repo sync # Pull the actual code
3) docker run -p $PWD:/src $container make -f /src/ROCm/tools/rocm,-build/ROCm.mk -j 50 # Do the build

It is not required to use dockerr, it is just a nice way to get a controlled build environment with all needed tools installed..

This is not to say that this way is better or worse than using spack, it is just different.

Reviewer 2: @amd-isparry - Wild. So, there's "TheRock", ROCm.mk makefile (docker-centric), and spack.

Reviewer 2: @garrettbyrd - do we call out specifically the benefit of being able to install select packages from ROCm for specific use cases, e.g., pytorch does not require all ROCm packages.

Author: @fluidnumerics-joe Yes, [this line](https://github.com/ROCm/rocm-blogs/blob/dfe33407a263ff5cf8e4605c28a70e59f53167d9/blogs/software-tools-optimization/spack-installation/README.md#using-spack-to-install-rocm) pushes that point.

> ... Although `mivisionx` comes packaged with ROCm by default, many ROCm components do not require this component. This underscores the utility of selective component installation via Spack.

## Install ROCm accelerated packages with Spack

The ROCm Documentation provides a [detailed guide on using Spack](https://rocm.docs.amd.com/projects/install-on-linux/en/docs-6.3.2/how-to/spack.html).

Getting started with Spack is quite easy. To install Spack, simply clone the repository and source `setup-env.sh`:

```sh
git clone https://github.com/spack/spack.git ~/spack/
source ~/spack/share/spack/setup-env.sh
```

You must also have C, C++, and Fortran compilers installed. To help Spack find your compilers, you can run the following

```sh
spack compiler find
```

To install `hipBLAS` targeting CDNA3 architectures, you can start by previewing what Spack will install by using `spack spec`

```sh
spack spec hipblas amdgpu_target=gfx942
```

<details><summary>Output of <code>spack spec hipblas amdgpu_target=gfx942</code></summary>

> ```

Reviewer 3: ```suggestion
```lang-shell
```

    install_tree: $HOME/opt/rocm
    view: $HOME/opt/rocm
```

It's worth noting that some packages related to ROCm that are available through `dnf`/`apt` are not available in Spack, e.g., `rocm-utils`. Also, some packages such as `aqlprofile` are not open source, and instead these are installed by grabbing the `.deb` from [repo.radeon.com](repo.radeon.com).

### Using Spack to install PyTorch with ROCm

Beyond ROCm, Spack also provides numerous Python packages, including PyTorch. Below is a `spack.yaml` that my be used to build PyTorch and its dependencies specifically for MI300X/MI300A.

`spack.yaml`:
```yaml
spack:
  specs:
  - py-torch+rocm
  concretizer:
    unify: true
  packages:
    all:
      prefer:
      - "amdgpu_target=gfx942"
  config:
    install_tree: $HOME/opt/rocm
    view: $HOME/opt/rocm
```

## Drivers

It is good to remind ourselves that the purpose of the ROCm software stack is to provide developers with the tools to program on (primarily) AMD GPUs. Under the hood, the developer still utilizes drivers to meaningfully interface with their AMD GPU and actually run programs. I.e., Spack and ROCm are used to build applications, drivers are used to run them.

> All drivers required for ROCm are fully open source at [`github.com/ROCm/ROCK-Kernel-Driver`](github.com/ROCm/ROCK-Kernel-Driver).

Author: Decided to keep the drivers section brief. I can come back and submit a quick PR to adjust this line (460) as needed.

Reviewer 3: THanks for the update.

---
blogpost: true
date: 26 Mar 2025
blog_title: "Installing ROCm from source with Spack"
author: 'Garrett Byrd, Joseph Schoonover'
tags: Scientific Computing, HPC, Installation
thumbnail: 'spack-thumbnail.jpg'
category: Software tools & optimizations
language: English
target_audience: AI/ML and HPC Developers.
key_value_propositions: Users can utilize the Spack package manager to easily install ROCm components from source.
myst:
    html_meta:
        "author": "Garrett Byrd, Joe Schoonover"

>         "description lang=en": "This blog post explains how to install ROCm and PyTorch from source using the Spack package manager."

Reviewer 4: This metadata field contains your blog's snippet text, which will appear on the ROCm Blogs portal as a teaser text to the post. The snippet text is capped at 150 characters. Please consider this revision:
```suggestion
        "description lang=en": "Install ROCm and PyTorch from source using Spack. Learn how to optimize builds, manage dependencies and streamline your GPU software stacks."
```

        "keywords": "Scientific Computing, HPC, PyTorch"
        "property=og:locale": "en_US"
        "amd_category": "Developer Resources"
        "amd_asset_type": "Blogs"
        "amd_blog_type": "Technical Articles & Blogs"
        "amd_technical_blog_type": "Tools, Features, and Optimizations"
        "amd_developer_type": "ML/AI Developer, Application Developer, HPC Developer"
        "amd_deployment": "Servers"
        "amd_product_type": "Development Tools, Software & Applications"
        "amd_developer_tool": "ROCm Software, Open-Source Tools"
        "amd_applications": "High Performance Computing"
        "amd_industries": "Data Center"
        "amd_blog_releasedate": Friday Mar 28, 12:00:00 PST 2025
---

# Installing ROCm from source with Spack

> This blog will walk through installing ROCm from source using the Spack package manager. We will also discuss Spack's place among other [ROCm installation methods](https://rocm.docs.amd.com/projects/install-on-linux/en/docs-6.3.2/install/install-overview.html), the landscape of [ROCm components](https://rocm.docs.amd.com/en/docs-6.3.2/what-is-rocm.html), and how ROCm, as an open-source software platform, allows developers to streamline software stacks for their applications.

Reviewer 4: Starting from the blog takeaway, and adding a more personal, call-to-action tone to the intro text:
```suggestion
In this guide you will learn how Spack makes building ROCm components from source easier and more flexible than other methods. The blog will walk you through installing ROCm from source using the Spack package manager. We will also discuss Spack's place among other [ROCm installation methods](https://rocm.docs.amd.com/projects/install-on-linux/en/docs-6.3.2/install/install-overview.html), the landscape of [ROCm components](https://rocm.docs.amd.com/en/docs-6.3.2/what-is-rocm.html), and show you how ROCm, as an open-source software platform, allows developers to streamline software stacks for their applications.
```

Author: Changed "The" to "This" but otherwise added this revision.

A *from-source* package manager is just that—a package manager that builds its packages from source code (when available). Compare this to a traditional package manager (e.g., `dpkg`/`apt` for Debian-based distros, or `yum`/`dnf` for RHEL-based distros) which installs pre-compiled binaries for each package. Put (extremely) simply, you can think of locally compiled binaries as being "fine-tuned" for the particular machine being used. Because Spack allows you to select the compilers and build optimizations when compiling source code, there is potential to achieve greater performance. Additionally, for GPU accelerated tools and libraries like those included in ROCm, building from source allows you to target specific GPU platforms rather than building for all supported architectures. This can result in reduction in storage costs for installing ROCm, which can be beneficial for creating lightweight container images that depend on ROCm. 

## Why install ROCm from source?

> If you are installing pre-compiled binaries on a supported operating system (e.g., [using `apt` to install ROCm on Ubuntu 24.04](https://rocm.docs.amd.com/projects/install-on-linux/en/docs-6.3.2/install/quick-start.html)), ROCm and its components come pre-built for [any ROCm-supported GPU](https://rocm.docs.amd.com/projects/install-on-linux/en/docs-6.3.2/reference/system-requirements.html). These builds are ideal for most linux package manager installs, so that everyday users need not worry about specific `gfx` versions.

Reviewer 4: Please capitalize Linux when needed throughout the blog
```suggestion
If you are installing pre-compiled binaries on a supported operating system (e.g., [using `apt` to install ROCm on Ubuntu 24.04](https://rocm.docs.amd.com/projects/install-on-linux/en/docs-6.3.2/install/quick-start.html)), ROCm and its components come pre-built for [any ROCm-supported GPU](https://rocm.docs.amd.com/projects/install-on-linux/en/docs-6.3.2/reference/system-requirements.html). These builds are ideal for most Linux package manager installs, so that everyday users need not worry about specific `gfx` versions.
```

When installed through a traditional package manager, every component is built against every (supported) `gfx` version. For example, [rocBLAS builds against fourteen different `gfx` versions](https://github.com/ROCm/ROCm/blob/29ba151b48c34fa2129a87097936200ab5b494d8/tools/rocm-build/build_rocblas.sh#L34) (including `xnack` variants). For example, if you are developing a ROCm-dependent application for a cluster of AMD Instinct™ MI300X Accelerators, you would only need ROCm components and kernels built against `gfx942`.

> There are are a few different methods by which you can install ROCm from source, such as [`make`](https://github.com/ROCm/ROCm/blob/develop/tools/rocm-build/README.md)  and [TheRock](https://github.com/ROCm/TheRock), AMD's lightweight open source build system for HIP and ROCm. The advantage of Spack is that it provides incredible ease-of-use. With each release of ROCm, AMD has consistently kept the corresponding Spack packages up to date. Continuing with our MI300X example, we can just `spack install hipblas` to install `hipBLAS`. (We can further influence the specifics of this build process, as we will see below.)

Reviewer 4: Typo
```suggestion
There are a few different methods by which you can install ROCm from source, such as [`make`](https://github.com/ROCm/ROCm/blob/develop/tools/rocm-build/README.md)  and [TheRock](https://github.com/ROCm/TheRock), AMD's lightweight open source build system for HIP and ROCm. The advantage of Spack is that it provides incredible ease-of-use. With each release of ROCm, AMD has consistently kept the corresponding Spack packages up to date. Continuing with our MI300X example, we can just `spack install hipblas` to install `hipBLAS`. (We can further influence the specifics of this build process, as we will see below.)
```

As an example, we can set up an environment using a yaml file that defines what packages we want to install, how to install them, and where to install them. Below is an example environment file that installs `hipblas` and all of its dependencies targeting the `gfx942` AMD GPU architecture.

```yaml
spack:
  specs:
  - hipblas@6.3.2
  concretizer:
    unify: true
  packages:
    all:
      prefer:
      - "amdgpu_target=gfx942"
  view: /opt/rocm
```

In this example, hipBLAS and all of its dependencies will be installed in a single Spack managed location and the filesystem view of the packages will be available at `/opt/rocm`. In essence, once this environment is installed, this would provide a hipBLAS installation in much the same way that `apt install hipblas` would.

To work with environments, suppose this example environment file is stored in `$HOME/spack-env/spack.yaml`. You can activate this environment and concretize the build with the following command:

```sh
spack env activate $HOME/spack-env
spack concretize -f
```

To install this environment, simply run:
```sh
spack install
spack gc -y
```

> The last step here performs "garbage collection", removing packages that are only build-time dependencies. This can help lighten up the final installation.

Reviewer 4: Consider adding a more detailed explanation, maybe along these lines:
```suggestion
The last step here performs "garbage collection", removing packages that are only build-time dependencies. This reduces image size by removing unused build-time dependencies and helps lighten up the final installation.
```

Author: I didn't use the exact suggested wording, but a revision was made.

Read more about Spack environments [here](https://spack.readthedocs.io/en/latest/environments.html).

### Using Spack environments to create Docker image recipes

> Another major benefit of using Spack environments is that you can easily create a Dockerfile to install just the ROCm packages you care about for the specific architectures you'd like to target. In continuing with our previous example, we can create a Dockerfile by doing the following :

Reviewer 4: Please consider adding a transition sentence here, briefly explaining the rationale for creating a Dockerfile from a Spack environment:
```suggestion
Another major benefit of using a Spack environment is that you can easily containerize it to ensure reproducibility and portability. To do that, you can create a Dockerfile to install just the ROCm packages you care about for the specific architectures you'd like to target. In continuing with our previous example, we can create a Dockerfile by doing the following :
```

ROCm officially supports roughly 70 components, but the exact list of component varies across multiple sources. Here are a few:

- [What is ROCm?](https://rocm.docs.amd.com/en/latest/what-is-rocm.html) $^{[1]}$
- [`ROCm/default.xml`](https://github.com/ROCm/ROCm/blob/develop/default.xml) $^{[2]}$
- [ROCm Documentation for Spack](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/how-to/spack.html#rocm-packages-in-spack) $^{[1]}$
- [ROCm Compatibility Matrix](https://rocm.docs.amd.com/en/latest/compatibility/compatibility-matrix.html) $^{[1]}$
- [Version release notes](https://rocm.docs.amd.com/en/latest/about/release-notes.html#rocm-components) $^{[1]}$

$^{[1]}$ Pages from ROCm Documentation

$^{[2]}$ ROCm GitHub repository

> The landscape of ROCm components is best laid out by the following graphic, taken from the [What is ROCm?](https://rocm.docs.amd.com/en/latest/what-is-rocm.html) page from the documentation.

Reviewer 4: ```suggestion
The landscape of ROCm components is best laid out by the following graphic (Figure 1, below), taken from the [What is ROCm?](https://rocm.docs.amd.com/en/latest/what-is-rocm.html) page from the documentation.
```

<img src="./images/rocm-software-stack-6_3_2.jpg">

<p style="text-align:center">
Figure 1: The ROCm software stack
</p>

The ROCm stack relies on lower layers of runtimes and compilers which are generally essential for most ROCm components; i.e., these are the "meat and potatoes" of ROCm. Above these layers are **Tools**, which are quite handy for developers, but are also used by some ROCm components. The ROCm stack becomes particularly discrete at the **Library** level. E.g., `rocJPEG` does not depend on `hipTensor` and vice versa. This motivates the question, "Why should we install all of ROCm?" if we just need a specific component, and drives us to understand specific dependencies within the ROCm stack.

### ROCm Component Dependencies

> As a package manager, Spack is also a great system for understanding the dependencies of the ROCm stack. The following graphic has been generated using the outputs of `spack spec <package>`.

Reviewer 4: ```suggestion
As a package manager, Spack is also a great system for understanding the dependencies of the ROCm stack. The following graphic (Figure 2, below) has been generated using the outputs of `spack spec <package>`.
```

### Using Spack to install PyTorch with ROCm

Beyond ROCm, Spack also provides numerous Python packages, including PyTorch. Below is a `spack.yaml` that my be used to build PyTorch and its dependencies specifically for MI300X/MI300A.

`spack.yaml`:
```yaml
spack:
  specs:
  - py-torch+rocm
  concretizer:
    unify: true
  packages:
    all:
      prefer:
      - "amdgpu_target=gfx942"
  config:
    install_tree: $HOME/opt/rocm
    view: $HOME/opt/rocm
```

## Drivers

It is good to remind ourselves that the purpose of the ROCm software stack is to provide developers with the tools to program on (primarily) AMD GPUs. Under the hood, the developer stills needs the AMDGPU drivers to meaningfully interface with their AMD GPU and actually run programs. I.e., Spack and ROCm are used to build applications, drivers are used to run them.

See: [Linux Drivers for AMD Radeon Graphics](https://www.amd.com/en/support/download/linux-drivers.html)

For developers on Radeon systems, these drivers are [fully open source](https://github.com/ROCm/ROCK-Kernel-Driver) and are [included in the Linux kernel](https://www.kernel.org/doc/html/latest/gpu/amdgpu/index.html).

> However, AMD also offers PRO drivers for Radeon PRO and Instinct cards. These drivers are not open source, but are included with ROCm. (PRO drivers are also [available for manual install](https://www.amd.com/en/support/download/linux-drivers.html).)

Reviewer 3: This is incorrect. Drivers for Instinct cards are also open sourced. The repo here contains the driver distributed with ROCm. https://github.com/ROCm/ROCK-Kernel-Driver. Your statement that this repo is for radeon users is incorrect. I understand that our driver story for ROCm is a little confusing. I actually have a blog explaining that will go live in the coming weeks. If you can point me via email (saad.rahim@ three letter acronym for the company I work for) to an NDA, I can share a preview.

PRO drivers should not be used for Instinct GPUs.

Reviewer 3: The driver statements here are the only thing I want to clear up prior to publishing. And the image if you want us to help you generate it.

Author: Thanks for the correction. I just sent an email your way.

## Summary

This blog provides a basic introduction to the Spack package manager and how to install ROCm components using Spack. Specific advantages of Spack have been discussed, such as

<!-- markdownlint-disable -->
- Easily building ROCm components from source.
- Installing Spack packages (such as PyTorch) targeting specific `gfx` architectures.
- Using Spack for cleaning up build dependencies.
- Utilizing `spack spec` to understand ROCm dependencies.
<!-- markdownlint-restore -->

This post also highlights the landscape of the ROCm ecosystem and provides an overview of standard and upcoming ROCm components.

## Acknowledgments

Special thanks to [Garrett Byrd](https://github.com/garrettbyrd) and [Dr. Joe Schoonover](https://github.com/fluidnumerics-joe) at [Fluid Numerics](https://www.fluidnumerics.com/) for contributing this blog. The ROCm software ecosystem is strengthened by community projects that enable you to use AMD GPUs in new ways. If you have a project you would like to share here, [please raise an issue or PR](https://github.com/ROCm/rocm-blogs).

> ### Find Fluid Numerics online at:

Reviewer 3: ```suggestion
### Find Fluid Numerics online
```
